keras/engine/topology.py

```diff
@@ -1314,13 +1334,40 @@ def get_output_shape_for(self, input_shape):
         elif self.mode == 'cos':
             return (input_shapes[0][0], 1)
 
-    def compute_mask(self, input, mask=None):
-        '''TODO: add mask merging support
-        '''
-        if mask is not None and any(mask):
-            raise Exception('Merge does not support masking, ' +
-                            'but was passed an input mask: ' + str(mask))
-        return None
+    def _normalize_mask_dims(self, mask, input):
+        if mask is None:
+            return K.ones_like(input)
+
+        if K.ndim(mask) == K.ndim(input):
+            return K.cast(mask, K.floatx())
+
+        assert K.ndim(mask) == K.ndim(input) - 1
+        mask = K.expand_dims(mask)
+        return K.ones_like(input) * K.cast(mask, K.floatx())
+
+    def compute_mask(self, inputs, mask=None):
+        if mask is None or not any([m is not None for m in mask]):
+            return None
+
+        assert hasattr(mask, '__len__')
+        assert len(mask) == len(inputs)
+
+        if self.mode in ['sum', 'mul', 'ave']:
+            normalized = [self._normalize_mask_dims(m, i) for m, i in zip(mask, inputs)]
+            return K.eall(normalized)
+        elif self.mode in ['concat']:
```

Comment:
If concatenation is not on the time axis, then masks have to be AND-merged on the time axis.